resolver-dns/src/main/java/io/netty/resolver/dns/DefaultDnsCache.java

```diff
@@ -289,4 +253,92 @@ public String toString() {
+    private static final class Entries extends AtomicReference<List<DefaultDnsCacheEntry>> {
+
+        Entries(DefaultDnsCacheEntry entry) {
+            set(Collections.singletonList(entry));
+        }
+
+        void add(DefaultDnsCacheEntry e) {
+            if (e.cause() == null) {
+                for (;;) {
+                    List<DefaultDnsCacheEntry> entries = get();
+                    if (!entries.isEmpty()) {
+                        final DefaultDnsCacheEntry firstEntry = entries.get(0);
+                        if (firstEntry.cause() != null) {
+                            assert entries.size() == 1;
+                            if (compareAndSet(entries, Collections.singletonList(e))) {
+                                firstEntry.cancelExpiration();
+                                return;
+                            } else {
+                                // Need to try again as CAS failed
+                                continue;
+                            }
+                        }
+                        // Create a new List for COW semantics
+                        List<DefaultDnsCacheEntry> newEntries = new ArrayList<DefaultDnsCacheEntry>(entries.size() + 1);
+                        newEntries.addAll(entries);
+                        newEntries.add(e);
+                        if (compareAndSet(entries, newEntries)) {
+                            return;
+                        }
+                    } else if (compareAndSet(entries, Collections.singletonList(e))) {
+                        return;
+                    }
+                }
+            } else {
+                List<DefaultDnsCacheEntry> entries = getAndSet(Collections.singletonList(e));
+                int numEntries = entries.size();
+                for (int i = 0; i < numEntries; i ++) {
+                    entries.get(i).cancelExpiration();
+                }
+            }
+        }
+
+        boolean remove(DefaultDnsCacheEntry entry) {
+            for (;;) {
+                List<DefaultDnsCacheEntry> entries = get();
+
+                if (entries.isEmpty()) {
+                    if (compareAndSet(entries, Collections.<DefaultDnsCacheEntry>emptyList())) {
+                        return false;
+                    }
+                } else {
+                    // Just size the new ArrayList as before as we may not find the entry we are looking for and not
+                    // want to cause an extra allocation / memory copy in this case.
+                    List<DefaultDnsCacheEntry> newEntries = new ArrayList<DefaultDnsCacheEntry>(entries.size());
+                    for (int i = 0; i < entries.size(); i++) {
+                        DefaultDnsCacheEntry e = entries.get(i);
+                        if (!e.equals(entry)) {
+                            newEntries.add(e);
+                        }
+                    }
+                    if (newEntries.isEmpty()) {
```

Comment:
would it be better to handle this by checking `if (entries.size() == 1 && entry.get(0).equals(newEntry)) { swap with empty collection }`? or is it possible that `entries` may contain multiple elements which equal `entry`?

Reply:
@Scottmitch it may be possible to have multiple entries so I think the way its handled now it more safe.

Reply:
`DefaultDnsCacheEntry` doesn't implement `equals`, and on each `cache(..)` call a new element is created. Can you clarify under which conditions the `entries` array would contain the same element multiple times?

Reply:
@Scottmitch actually scratch this... you are right it cant happen.

Reply:
Now that we have a special check for `size == 1` before this conditional block executes ... is this ever possible?

Reply:
no... let me remove it.